File: src/agent.ts
import Anthropic from "@anthropic-ai/sdk";
import fs from "fs";
import path from "path";
import { BigQueryClient } from "./bigquery-client.js";
import { generateXlsx } from "./exporter.js";
import { enrichQuestion, buildUserContext, type UserInfo } from "./prompt-enricher.js";
import { TOOLS } from "./tools.js";
import { config } from "./config.js";

export class InsightsAgent {
  private claude: Anthropic;
  private bq: BigQueryClient;
  private user: UserInfo | null;
  private systemPrompt: string;
  private conversation: Anthropic.MessageParam[] = [];

  constructor(user: UserInfo | null = null) {
    this.claude = new Anthropic();
    this.bq = new BigQueryClient();
    this.user = user;
    this.systemPrompt = "";
  }

  async init(): Promise<void> {
    const template = fs.readFileSync(path.resolve("prompts/system.txt"), "utf-8");
    let schema: string;
    try {
      schema = await this.bq.getFullSchema();
    } catch {
      schema = "(Schema será descoberto dinamicamente via tools)";
    }
    const userContext = buildUserContext(this.user);
    const today = new Date().toLocaleDateString("pt-BR");

    this.systemPrompt = template
      .replace("{schema}", schema)
      .replace("{user_context}", userContext)
      .replace("{today}", today);
  }

  private async executeTool(name: string, input: Record<string, unknown>): Promise<string> {
    switch (name) {
      case "execute_query":
        return this.bq.executeQuery(input.sql as string, input.max_rows as number | undefined);
      case "list_datasets":
        return this.bq.listDatasets();
      case "list_tables":
        return this.bq.listTables(input.dataset_id as string);
      case "describe_table":
        return this.bq.describeTable(input.dataset_id as string, input.table_id as string);
      case "export_spreadsheet":
        try {
          const rows = await this.bq.executeQueryRaw(input.sql as string, (input.max_rows as number) || 5000);
          const fileId = await generateXlsx(rows, (input.title as string) || "Relatório");
          return `EXPORT_FILE_ID:${fileId}|ROWS:${rows.length}|TITLE:${input.title || "Relatório"}`;
        } catch (e) {
          return `ERRO ao exportar: ${(e as Error).message}`;
        }
      default:
        return `Tool desconhecida: ${name}`;
    }
  }

  async ask(question: string): Promise<string> {
    if (!this.systemPrompt) await this.init();

    const enriched = enrichQuestion(question);
    this.conversation.push({ role: "user", content: enriched });

    for (let i = 0; i < 10; i++) {
      const response = await this.claude.messages.create({
        model: config.claudeModel,
        max_tokens: config.claudeMaxTokens,
        system: this.systemPrompt,
        tools: TOOLS,
        messages: this.conversation,
      });

      if (response.stop_reason === "tool_use") {
        const toolResults: Anthropic.ToolResultBlockParam[] = [];
        for (const block of response.content) {
          if (block.type === "tool_use") {
            const result = await this.executeTool(block.name, block.input as Record<string, unknown>);
            toolResults.push({ type: "tool_result", tool_use_id: block.id, content: result });
          }
        }
        this.conversation.push({ role: "assistant", content: response.content });
        this.conversation.push({ role: "user", content: toolResults });
      } else if (response.stop_reason === "end_turn") {
        this.conversation.push({ role: "assistant", content: response.content });
        return response.content
          .filter((b): b is Anthropic.TextBlock => b.type === "text")
          .map((b) => b.text)
          .join("");
      }
    }

    return "Limite de iterações atingido. Tente reformular a pergunta.";
  }

  reset(): void {
    this.conversation = [];
  }
}

File: src/agent.py
import anthropic
from pathlib import Path
from src.bigquery_client import BigQueryClient
from src.tools import TOOLS
from config.settings import CLAUDE_MODEL, CLAUDE_MAX_TOKENS


class InsightsAgent:
    def __init__(self):
        self.claude = anthropic.Anthropic()
        self.bq = BigQueryClient()
        self.system_prompt = self._load_system_prompt()
        self.conversation: list[dict] = []

    def _load_system_prompt(self) -> str:
        template = Path("prompts/system.txt").read_text()
        try:
            schema = self.bq.get_full_schema()
        except Exception:
            schema = "(Schema será descoberto dinamicamente via tools list_datasets/list_tables/describe_table)"
        return template.replace("{schema}", schema)

    def _execute_tool(self, name: str, inputs: dict) -> str:
        match name:
            case "execute_query":
                return self.bq.execute_query(
                    sql=inputs["sql"],
                    max_rows=inputs.get("max_rows", 100),
                )
            case "list_datasets":
                return self.bq.list_datasets()
            case "list_tables":
                return self.bq.list_tables(inputs["dataset_id"])
            case "describe_table":
                return self.bq.describe_table(inputs["dataset_id"], inputs["table_id"])
            case _:
                return f"Tool desconhecida: {name}"

    def ask(self, question: str) -> str:
        self.conversation.append({"role": "user", "content": question})

        max_iterations = 10  # evitar loops infinitos
        for _ in range(max_iterations):
            response = self.claude.messages.create(
                model=CLAUDE_MODEL,
                max_tokens=CLAUDE_MAX_TOKENS,
                system=self.system_prompt,
                tools=TOOLS,
                messages=self.conversation,
            )

            if response.stop_reason == "tool_use":
                # Executar todas as tool calls
                tool_results = []
                for block in response.content:
                    if block.type == "tool_use":
                        result = self._execute_tool(block.name, block.input)
                        tool_results.append({
                            "type": "tool_result",
                            "tool_use_id": block.id,
                            "content": result,
                        })

                self.conversation.append({"role": "assistant", "content": response.content})
                self.conversation.append({"role": "user", "content": tool_results})

            elif response.stop_reason == "end_turn":
                self.conversation.append({"role": "assistant", "content": response.content})
                return "".join(
                    block.text for block in response.content if hasattr(block, "text")
                )

        return "Limite de iterações atingido. Tente reformular a pergunta."

    def reset(self):
        self.conversation = []
